File: src/mangadex.rs
extern crate reqwest;
extern crate serde;

use serde::Deserialize;
use serde::Serialize;
use serde_json::Value as JsonValue;

use std::io::{self, Write};

use indicatif::ProgressBar;

//-------------------------------------------------------------------------------------------------

type MangaData = MangaDataJSONResponse;

#[derive(Debug, Deserialize, Serialize)]
struct MangaDataJSONResponse {
  result: String,
  data: Data,
}

#[derive(Debug, Deserialize, Serialize)]
struct Data {
  id: String,
  attributes: MangaAttributes,
}

#[derive(Debug, Deserialize, Serialize)]
struct MangaAttributes {
  title: Title,
  availableTranslatedLanguages: Vec<String>,
}

#[derive(Debug, Deserialize, Serialize)]
struct Title {
  en: String,
}

fn getManga(_mangaId: String) -> MangaData {

  let baseUrl = format!("https://api.mangadex.org/manga/{}", _mangaId);
  let url = reqwest::Url::parse(&baseUrl).unwrap();

  let json: JsonValue = reqwest::blocking::get(url).expect("bad request").json().expect("error parsing json");

  let mangaData: MangaData = serde_json::from_str(&json.to_string()).unwrap();

  // println!("mangaData = {:?}", mangaData.data.id);

  return mangaData;
}

//-------------------------------------------------------------------------------------------------

type MangaChapters = MangaChaptersJSONResponse;

#[derive(Debug, Deserialize, Serialize)]
struct MangaChaptersJSONResponse {
  result: String,
  data: Vec<ChapterData>,
  total: i32,
}

#[derive(Debug, Deserialize, Serialize)]
struct ChapterData {
  id: String,
  attributes: ChapterAttributes,
}

#[derive(Debug, Deserialize, Serialize)]
struct ChapterAttributes {
  chapter: Option<String>,
  title: Option<String>,
}

fn getMangaChapters(_mangaInfo: &MangaData) -> MangaChapters {
  // println!("_mangaInfo.data.id = {:?}", _mangaInfo.data.id);
  std::thread::sleep(std::time::Duration::from_millis(3000)); // debugger for directory versions

  let queryLimit: i32 = 500;
  let mut offset: i32 = 0;
  let selectedLanguage: String = "en".to_string();
  let baseUrl = format!("https://api.mangadex.org/manga/{}/feed?includeFuturePublishAt=0&limit={}&offset={}&translatedLanguage[]={}", _mangaInfo.data.id, queryLimit, offset, selectedLanguage);

  let url = reqwest::Url::parse(&baseUrl).unwrap();

  let json: JsonValue = reqwest::blocking::get(url).expect("bad request").json().expect("error parsing json");

  let mut mangaChapters: MangaChapters = serde_json::from_str(&json.to_string()).unwrap();

  if mangaChapters.total > queryLimit {
    std::thread::sleep(std::time::Duration::from_millis(3000)); // debugger for directory versions
    offset = 500;
    let baseUrl = format!("https://api.mangadex.org/manga/{}/feed?includeFuturePublishAt=0&limit={}&offset={}&translatedLanguage[]={}", _mangaInfo.data.id, queryLimit, offset, selectedLanguage);
    let url = reqwest::Url::parse(&baseUrl).unwrap();
    let json: JsonValue = reqwest::blocking::get(url).expect("bad request").json().expect("error parsing json");
    let parsed: MangaChapters = serde_json::from_str(&json.to_string()).unwrap();
    mangaChapters.data.extend(parsed.data.into_iter());
  }

  // println!("mangaChapters = {:?}", mangaChapters);
  return mangaChapters;
}

//-------------------------------------------------------------------------------------------------

type MangaImages = MangaImagesJSONResponse;

#[derive(Debug, Deserialize, Serialize)]
struct MangaImagesJSONResponse {
  result: String,
  chapter: ChapterImages,
}

#[derive(Debug, Deserialize, Serialize)]
struct ChapterImages {
  hash: String,
  data: Vec<String>,
}

fn getMangaChapterImages(_mangaTitle: String, _mangaChapters: &MangaChapters, _userInput: String, _singleFolder: bool) {
  let mut userInputVec: Vec<_> = [].to_vec();
  let progressBarLength = if !_userInput.trim().eq("") {
    let chapterSeletion = _userInput.trim().split(" ").map(|chapter| chapter.to_string());
    userInputVec = chapterSeletion.collect();
    userInputVec.len()
  } else {
    _mangaChapters.data.len()
  };

  println!("\nDownloading");
  // let progressBar = ProgressBar::new(progressBarLength.try_into().unwrap());
  // progressBar.inc(0);

  let mangaTitleDirectory = format!("downloads/{}", _mangaTitle);
  std::fs::create_dir_all(mangaTitleDirectory).unwrap();

  let mut i: usize = 0;
  let mut k: usize = 0;
  'i: loop {
    let baseUrl = format!("https://api.mangadex.org/at-home/server/{}", _mangaChapters.data[i].id);

    let url = reqwest::Url::parse(&baseUrl).unwrap();

    let json: JsonValue = reqwest::blocking::get(url).expect("bad request").json().expect("error parsing json");

    let mangaChapterImages: MangaImages = serde_json::from_str(&json.to_string()).unwrap();

    let hash: String = mangaChapterImages.chapter.hash;
    let chapterImagesFileName: Vec<String> = mangaChapterImages.chapter.data;

    if _userInput.trim().eq("oneshot") || userInputVec.iter().any(|k| k.eq(&_mangaChapters.data[i].attributes.chapter.as_ref().expect("expect chapter not to be null").to_string())) || _userInput.trim().eq("") {
      let directory = if _singleFolder {
        // let directory = if _singleFolder && chapterImagesFileName.len() == 1 {
        // if chapterImagesFileName.len() == 1 {
          format!("downloads/{}/", _mangaTitle)
        // } else {
        //   format!("downloads/{}/Ch.{} - {}/", _mangaTitle, _mangaChapters.data[i].attributes.chapter.as_ref().expect("expect chapter not to be null").to_string(), _mangaChapters.data[i].attributes.title.as_ref().expect("expect title not to be null").to_string())
        // }
      } else if _userInput.trim().eq("oneshot") {
        format!("downloads/{}/Oneshot/", _mangaTitle)
      } else {
        match &_mangaChapters.data[i].attributes.title {
          Some(_) => format!("downloads/{}/Ch.{} - {}/", _mangaTitle, _mangaChapters.data[i].attributes.chapter.as_ref().expect("expect chapter not to be null").to_string(), _mangaChapters.data[i].attributes.title.as_ref().expect("expect title not to be null").to_string()),
          None => format!("downloads/{}/Ch.{}/", _mangaTitle, _mangaChapters.data[i].attributes.chapter.as_ref().expect("expect chapter not to be null").to_string()),
          // None => format!("downloads/{}/Ch.{}/", _mangaTitle, "null chapter"), // if this is null, it's probably a oneshot
        }
      };
      let mut dirVersion = 2;
      let mut stringDir = "".to_string();
      stringDir.push_str(&directory);
      stringDir.pop();
      stringDir.push_str(" - V");
      loop {
        if _singleFolder {
          break;
        }
        if std::fs::metadata(&directory).is_ok() {
          stringDir.push_str(&dirVersion.to_string());
          if !std::fs::metadata(&stringDir).is_ok() {
            std::fs::create_dir_all(&stringDir).unwrap();
            break;
          } else {
            dirVersion += 1;
          }
        } else {
          std::fs::create_dir_all(&directory).unwrap();
          break;
        }
      }
///////////////////////////////////////////////////////////////////////////////////////////////////
      let mut j: usize = 0;
      'j: loop {
        let fileExtension = if chapterImagesFileName[j].contains(".jpg") {
          "jpg"
        } else if chapterImagesFileName[j].contains(".jpeg") {
          "jpeg"
        } else {
          "png"
        };

        let mut fileName = if _singleFolder {
          if chapterImagesFileName.len() == 1 {
            format!("{}/{}.{}", &directory, _mangaChapters.data[i].attributes.chapter.as_ref().expect("expect title not to be null").to_string(), fileExtension)
          } else {
            if _userInput.trim().eq("oneshot") {
              format!("{}/{}.{}", &directory, j + 1, fileExtension) // change this to just increment a number by the side of the chapter number
            } else {
              format!("{}/{} - {}.{}", &directory, _mangaChapters.data[i].attributes.chapter.as_ref().expect("expect title not to be null").to_string(), j + 1, fileExtension)
            }
          }
        } else {
          if j + 1 < 10 {
            format!("{}/00{}.{}", &directory, j + 1, fileExtension)
          } else {
            format!("{}/0{}.{}", &directory, j + 1, fileExtension)
          }
        };

        if _singleFolder && std::path::Path::new(&fileName).exists() {
          let mut chapterVersion = 2;
          let mut stringFile = "".to_string();
          stringFile.push_str(&fileName);
          // let charIndex = stringFile.chars().position(|c| c == '.').unwrap();
          let charIndex = stringFile.chars().count() - stringFile.chars().rev().position(|c| c == '.').unwrap() - 1;
          // let charIndex = stringFile.rfind('.').unwrap() - 1;
          loop {
            // if !_singleFolder {
            //   break;
            // }
            let mut auxFileName = "".to_string();
            auxFileName.push_str(&stringFile);
            if std::path::Path::new(&auxFileName).exists() {
              auxFileName.as_bytes()[charIndex - 1].to_string().push_str(&format!(" - V{}", chapterVersion).to_string());
              if !std::path::Path::new(&auxFileName).exists() {
                fileName = auxFileName;
                break;
              } else {
                chapterVersion += 1;
              }
            }
          }
        }
        print!("fileName {}\n", fileName);
        // let mut file = std::fs::File::create(fileName).unwrap(); // TODO - if file exists and _singleFolder, add a v2 at the end
        // let baseUrl = format!("https://uploads.mangadex.org/data/{}/{}", hash, chapterImagesFileName[j]);
        // let url = reqwest::Url::parse(&baseUrl).unwrap();
        // let _mangaImage = reqwest::blocking::get(url).unwrap().copy_to(&mut file).unwrap();

        if j < chapterImagesFileName.len() - 1 {
          j += 1;
        } else {
          break 'j;
        }
      }
      std::thread::sleep(std::time::Duration::from_millis(2000)); // debugger for directory versions
      // if k < progressBarLength - 1 {
      //   k += 1;
      //   progressBar.inc(1);
      // } else {
      //   progressBar.finish_with_message("done");
      //   break 'i;
      // }
    }
    i += 1;
  }
}

//-------------------------------------------------------------------------------------------------

pub fn mangadex(_mangaId: String) {
  let mangaInfo = getManga(_mangaId);
  if mangaInfo.result != "ok" {
    print!("error fecthing manga. error: {}", mangaInfo.result);
    return;
  }
  let mangaChapters = getMangaChapters(&mangaInfo);
  let mangaTitle = mangaInfo.data.attributes.title.en;
  print!("{}\n", mangaTitle);
  if mangaChapters.data.len() == 0 {
    print!("No chapters available\n");
    return;
  }
  print!("Number of chapters: {}\n\n", mangaChapters.data.len());

  let mut arr: Vec<_> = Vec::<_>::new();
  print!("available chapters:\n");
  if mangaChapters.data.len() == 1 && mangaChapters.data[0].attributes.chapter.as_ref().is_none() {
    print!("Oneshot\n");
  } else {
    for chapter in mangaChapters.data.iter() {
      if chapter.attributes.chapter.as_ref().is_some() {
        arr.push(chapter.attributes.chapter.as_ref().expect("expect title not to be null").to_string().parse::<f32>().unwrap());
      }
    }
    arr.sort_by(|a, b| a.partial_cmp(b).unwrap());
    for i in arr.iter() {
      print!("{}\n", i);
    }
  }

  print!("\nEnter the chapters you want to download\n");
  print!("Options: 'all', 'asf (all chapters in a single folder)', 'chapter numbers separated by spaces', 'oneshot', 'quit'\n");
  loop {
    print!("-> ");
    std::io::stdout().flush().expect("failed to flush stdout");

    let mut userInput = String::new();
    let stdin = io::stdin();
    stdin.read_line(&mut userInput).expect("Could not read line");

    match userInput.trim() {
      "all" => getMangaChapterImages(mangaTitle.to_string(), &mangaChapters, "".to_string(), false),
      "asf" => getMangaChapterImages(mangaTitle.to_string(), &mangaChapters, "".to_string(), true),
      "oneshot" => getMangaChapterImages(mangaTitle.to_string(), &mangaChapters, "oneshot".to_string(), true),
      "quit" => break,
      _ => getMangaChapterImages(mangaTitle.to_string(), &mangaChapters, userInput, false),
    }
    println!("\nDownload completed!\n");
    break;
  }
}

// TODO
// directory version
// oneshot
// 'asf' (tomo-chan alike. single page per chapter, so it saves everything in one directory)

// 192aa767-2479-42c1-9780-8d65a2efd36a  // Gachiakuta
// 76ee7069-23b4-493c-bc44-34ccbf3051a8  // Tomo-chan
// eb0494de-3b43-4d52-a808-63429c4a4239  // ako to bambi
// ead4b388-cf7f-448c-aec6-bf733968162c  // Hanabi - oneshot

File: src/mangadex.go
package src

import (
  "fmt"
  "errors"
  "io/ioutil"
  "net/http"
  "encoding/json"
  "sort"
  "strconv"
  "strings"
  "bufio"
  "os"
)

func mangadex() DownloadResult {
  fmt.Printf("\nEnter the Manga ID: ")
  var userInput string
  fmt.Scanf("%s", &userInput)

  mangaInfo, err := getMangaMangadex(userInput)
  if err != nil {
    fmt.Println(err)
    return downloadFailed("", err.Error())
  }
  if mangaInfo.Result != "ok" {
    fmt.Println(mangaInfo.Result)
    return downloadFailed("", mangaInfo.Result)
  }

  mangaChapters, err := getMangaChaptersMangadex(mangaInfo)
  if err != nil {
    fmt.Println(err)
    return downloadFailed("", err.Error())
  }
  if len(mangaChapters.Data) == 0 {
    fmt.Println("No chapters available\n")
    // return
  }
  var mangaTitle string
  if mangaInfo.Data.Attributes.Title.EN != "" {
    mangaTitle = mangaInfo.Data.Attributes.Title.EN
  } else if mangaInfo.Data.Attributes.Title.JARomaji != "" {
    mangaTitle = mangaInfo.Data.Attributes.Title.JARomaji
  } else if mangaInfo.Data.Attributes.Title.JA != "" {
    mangaTitle = mangaInfo.Data.Attributes.Title.JA
  } else {
    mangaTitle = "manga - unknown title"
  }
  fmt.Println("")
  fmt.Println(mangaTitle)
  fmt.Println("")
  fmt.Println("Number of chapters: ", len(mangaChapters.Data))
  fmt.Println("Available chapters:")

  var arr []string
  if len(mangaChapters.Data) == 1 && mangaChapters.Data[0].Attributes.Chapter == nil {
    fmt.Println("Oneshot")
  } else {
    for _, chapter := range mangaChapters.Data {
      if chapter.Attributes.Chapter != nil {
        arr = append(arr, *chapter.Attributes.Chapter)
      }
    }

    for _, i := range arr {
      fmt.Println(i)
    }
  }

  fmt.Println("\nEnter the chapters you want to download\n")
  fmt.Println("Options: 'all', 'asf (all chapters in a single folder)', 'range', 'chapter numbers separated by spaces', 'oneshot', 'covers', 'quit'\n")
  // One reader for the whole prompt - a fresh one per read would buffer ahead
  // and swallow whatever the range prompts are waiting on
  _input := bufio.NewReader(os.Stdin)
  cancelled := false
  // Only the range option knows how many chapters it picked
  selectedChapters := 0
  loop: for {
    fmt.Printf("-> ")
    userInput, _ := _input.ReadString('\n')
    userInput = strings.TrimSuffix(userInput, "\n")

    switch userInput {
      case "all":
        getMangaChapterImagesMangadex(mangaTitle, mangaChapters, "", false)
        break loop
      case "asf":
        getMangaChapterImagesMangadex(mangaTitle, mangaChapters, "", true)
        break loop
      case "range":
        rangeChapters, err := selectChapterRangeMangadex(_input, mangaChapters)
        if err != nil {
          fmt.Println("\n" + err.Error())
          return downloadFailed(mangaTitle, err.Error())
        }
        selectedChapters = len(rangeChapters.Data)
        getMangaChapterImagesMangadex(mangaTitle, rangeChapters, "", false)
        break loop
      case "oneshot":
        getMangaChapterImagesMangadex(mangaTitle, mangaChapters, "oneshot", true)
        break loop
      case "covers":
        getMangaCoversMangadex(mangaTitle, mangaInfo.Data.ID)
        break loop
      case "quit":
        cancelled = true
        break loop
      default:
        getMangaChapterImagesMangadex(mangaTitle, mangaChapters, userInput, false)
    }
    break loop
  }

  if cancelled {
    return downloadCancelled()
  }

  fmt.Printf("\nDownload completed!\n")

  // Mangadex downloads every selected chapter inside one call, so outside of a
  // range there is no per chapter count to report here
  return downloadSuccess(mangaTitle, selectedChapters)
}

// The listing prints chapter numbers instead of positions, so the range is read
// as chapter numbers too - the same values the "chapter numbers" option takes.
// Decimals such as 10.5 are part of the numbering here, hence the float parse.
func selectChapterRangeMangadex(_input *bufio.Reader, _mangaChapters MangaChaptersMangadex) (MangaChaptersMangadex, error) {
  fmt.Println("\nEnter the range of chapters you want to download.")

  fmt.Printf("\nInitial chapter: ")
  userInputFirstChapter, _ := _input.ReadString('\n')
  firstChapter, err := strconv.ParseFloat(strings.TrimSpace(userInputFirstChapter), 64)
  if err != nil {
    return MangaChaptersMangadex{}, errors.New("Invalid initial chapter.")
  }

  fmt.Printf("\nLast chapter: ")
  userInputLastChapter, _ := _input.ReadString('\n')
  lastChapter, err := strconv.ParseFloat(strings.TrimSpace(userInputLastChapter), 64)
  if err != nil {
    return MangaChaptersMangadex{}, errors.New("Invalid last chapter.")
  }

  if lastChapter < firstChapter {
    firstChapter, lastChapter = lastChapter, firstChapter
  }

  // Same struct minus the chapters outside the range, so the download path can
  // treat it as if it were the whole feed
  rangeChapters := _mangaChapters
  rangeChapters.Data = nil
  for _, chapter := range _mangaChapters.Data {
    // Chapters with no number - oneshots and the like - cannot be placed in a
    // range, so they are left to the 'oneshot' option
    if chapter.Attributes.Chapter == nil {
      continue
    }
    chapterNumber, err := strconv.ParseFloat(*chapter.Attributes.Chapter, 64)
    if err != nil {
      continue
    }
    if chapterNumber >= firstChapter && chapterNumber <= lastChapter {
      rangeChapters.Data = append(rangeChapters.Data, chapter)
    }
  }
  rangeChapters.Total = len(rangeChapters.Data)

  if len(rangeChapters.Data) == 0 {
    return MangaChaptersMangadex{}, errors.New("No chapters in the selected range.")
  }

  fmt.Printf("\n")

  return rangeChapters, nil
}

type MangaDataMangadex struct {
  Result string `json:"result"`
  Data struct {
    ID string `json:"id"`
    Attributes struct {
      Title struct {
        EN string `json:"en"`
        JARomaji string `json:"ja-ro"`
        JA string `json:"ja"`
      } `json:"title"`
      AvailableTranslatedLanguages []string `json:"availableTranslatedLanguages"`
    } `json:"attributes"`
  } `json:"data"`
}

func getMangaMangadex(_mangaId string) (MangaDataMangadex, error) {
  var url string = fmt.Sprintf("https://api.mangadex.org/manga/%s", _mangaId)
  var mangaData MangaDataMangadex

  resp, err := http.Get(url)
  if err != nil {
    return mangaData, errors.New("Could not get manga info")
  }
  defer resp.Body.Close()
  body, err := ioutil.ReadAll(resp.Body)
  if err != nil {
    return mangaData, errors.New("Could not parse body")
  }
  if err := json.Unmarshal(body, &mangaData); err != nil {
    fmt.Println("Could not unmarshal JSON")
    return mangaData, errors.New("Could not unmarshal JSON")
  }
  return mangaData, nil
}

type MangaCoversDataMangadex struct {
  Result string `json:"result"`
  Data []struct {
    Attributes struct {
      Volume string `json:"volume"`
      FileName string `json:"fileName"`
      Locale string `json:"locale"`
    } `json:"attributes"`
  } `json:"data"`
  Total int `json:"total"`
}

func getMangaCoversMangadex(_mangaTitle string, _mangaId string) {
  // TODO
  // - this breaks on one piece - 100+ covers
  // - automate saving covers in the first chapter of their respective volume
  var _limit = 100
  var url string = fmt.Sprintf("https://api.mangadex.org/cover?limit=%d&manga[]=%s&order[createdAt]=asc&order[updatedAt]=asc&order[volume]=asc", _limit, _mangaId)
  var mangaCoversData MangaCoversDataMangadex

  resp, err := http.Get(url)
  if err != nil {
    fmt.Println("Could not get manga covers info")
  }
  defer resp.Body.Close()
  body, err := ioutil.ReadAll(resp.Body)
  if err != nil {
    fmt.Println("Could not parse body")
  }
  if err := json.Unmarshal(body, &mangaCoversData); err != nil {
    fmt.Println("Could not unmarshal JSON")
  }
  var dir string = fmt.Sprintf("%s%s/%s", downloadsRoot, authorSubDir, _mangaTitle)
  _dir := fsCreateDir(dir, true)
  for _, cover := range mangaCoversData.Data {
    // skip covers that are not en or jp
    if cover.Attributes.Locale == "ja" || cover.Attributes.Locale == "en" {
      var url string = fmt.Sprintf("https://uploads.mangadex.org/covers/%s/%s", _mangaId, cover.Attributes.FileName)
      var coverImage []byte
      for {
        resp, err := http.Get(url)
        if err != nil {
          fmt.Println("Request error. Retrying.")
          continue
        }
        defer resp.Body.Close()
        res, err := ioutil.ReadAll(resp.Body)
        if err != nil {
          fmt.Println("Request error. Retrying.")
        } else {
          coverImage = res
          break
        }
      }

      var coverFileName = fmt.Sprintf("Cover %s - %s", cover.Attributes.Volume, cover.Attributes.Locale)
      fsCreateFile(cover.Attributes.FileName, _dir, 0, coverImage, true, coverFileName)
    }
  }
}

type MangaChaptersMangadex struct {
  Result string `json:"result"`
  Data []struct {
    ID string `json:"id"`
    Attributes struct {
      Chapter *string `json:"chapter"`
      Title *string `json:"title"`
    } `json:"attributes"`
  } `json:"data"`
  Total int `json:"total"`
}

func getMangaChaptersMangadex(_mangaInfo MangaDataMangadex) (MangaChaptersMangadex, error) {
  var queryLimit int = 500
  var offset int = 0
  var selectedLanguage string = "en"

  var url string = fmt.Sprintf("https://api.mangadex.org/manga/%s/feed?includeFuturePublishAt=0&limit=%d&offset=%d&contentRating[]=safe&contentRating[]=suggestive&contentRating[]=erotica&contentRating[]=pornographic&translatedLanguage[]=%s", _mangaInfo.Data.ID, queryLimit, offset, selectedLanguage)
  var mangaChapters MangaChaptersMangadex

  resp, err := http.Get(url)
  if err != nil {
    return mangaChapters, errors.New("Could get manga chapter")
  }
  defer resp.Body.Close()
  body, err := ioutil.ReadAll(resp.Body)
  if err != nil {
    return mangaChapters, errors.New("Could not parse body")
  }
  if err := json.Unmarshal(body, &mangaChapters); err != nil {
    fmt.Println("Could not unmarshal JSON")
    return mangaChapters, errors.New("Could not unmarshal JSON")
  }

  if mangaChapters.Total > queryLimit {
    for offset < queryLimit {
      offset += 500
      var url string = fmt.Sprintf("https://api.mangadex.org/manga/%s/feed?includeFuturePublishAt=0&limit=%d&offset=%d&contentRating[]=safe&contentRating[]=suggestive&contentRating[]=erotica&contentRating[]=pornographic&translatedLanguage[]=%s", _mangaInfo.Data.ID, queryLimit, offset, selectedLanguage)
      var mangaChapters2 MangaChaptersMangadex

      resp, err := http.Get(url)
      if err != nil {
        return mangaChapters, errors.New("Could get manga chapters")
      }
      defer resp.Body.Close()
      body, err := ioutil.ReadAll(resp.Body)
      if err != nil {
        return mangaChapters, errors.New("Could not parse body")
      }
      if err := json.Unmarshal(body, &mangaChapters2); err != nil {
        fmt.Println("Could not unmarshal JSON")
        return mangaChapters, errors.New("Could not unmarshal JSON")
      }
      mangaChapters.Data = append(mangaChapters.Data, mangaChapters2.Data...)
    }
  }

  sort.Slice(mangaChapters.Data, func(i, j int) bool {
    var chapterA, chapterB string

    if mangaChapters.Data[i].Attributes.Chapter != nil {
      chapterA = *mangaChapters.Data[i].Attributes.Chapter
    }
    if mangaChapters.Data[j].Attributes.Chapter != nil {
      chapterB = *mangaChapters.Data[j].Attributes.Chapter
    }

    numA, _ := strconv.ParseFloat(chapterA, 64)
    numB, _ := strconv.ParseFloat(chapterB, 64)

    return numA < numB
  })

  return mangaChapters, nil
}

type MangaImagesMangadex struct {
  Result string `json:"result"`
  Chapter struct {
    Hash string `json:"hash"`
    Data []string `json:"data"`
  } `json:"chapter"`
}

func getMangaChapterImagesMangadex(_mangaTitle string, _mangaChapters MangaChaptersMangadex, _userInput string, _singleFolder bool) {
  fmt.Println("\nStarting Download")

  userInput := strings.Split(_userInput, " ")

  var i int = 0
  i: for {
    var chapterNameNoNIL string
    if _userInput == "oneshot" {
      chapterNameNoNIL = "Oneshot"
    } else if _mangaChapters.Data[i].Attributes.Chapter == nil {
      chapterNameNoNIL = ""
    } else {
      chapterNameNoNIL = *_mangaChapters.Data[i].Attributes.Chapter
    }
    if (contains(userInput, chapterNameNoNIL, _userInput == "oneshot")) || (_userInput == "oneshot") || (_userInput == "") {
      var url string = fmt.Sprintf("https://api.mangadex.org/at-home/server/%s", _mangaChapters.Data[i].ID)
      var mangaChapterImages MangaImagesMangadex

      for {
        resp, err := http.Get(url)
        if err != nil {
          fmt.Println("Request error. Retrying.")
          continue
        }
        defer resp.Body.Close()
        body, err := ioutil.ReadAll(resp.Body)
        if err != nil {
          fmt.Println("Request error. Retrying.")
        }
        if err := json.Unmarshal(body, &mangaChapterImages); err != nil {
          fmt.Println("Could not unmarshal JSON - manga chapter images")
        } else {
          break
        }
      }

      if len(mangaChapterImages.Chapter.Data) > 0 {
        // TODO - if "all" option, and name == nil then it is probably a oneshot. fix empty chapter number and set it to 0?
        var dir string
        if _singleFolder {
          dir = fmt.Sprintf("%s%s/%s", downloadsRoot, authorSubDir, _mangaTitle)
        } else if _userInput == "oneshot" {
          dir = fmt.Sprintf("%s%s/%s/Oneshot", downloadsRoot, authorSubDir, _mangaTitle)
        } else {
          if _mangaChapters.Data[i].Attributes.Title != nil && len(*_mangaChapters.Data[i].Attributes.Title) > 0 {
            dir = fmt.Sprintf("%s%s/%s/Ch.%s - %s", downloadsRoot, authorSubDir, _mangaTitle, chapterNameNoNIL, *_mangaChapters.Data[i].Attributes.Title)
          } else {
            dir = fmt.Sprintf("%s%s/%s/Ch.%s", downloadsRoot, authorSubDir, _mangaTitle, chapterNameNoNIL)
          }
        }
        fmt.Println("Downloading chapter: ", chapterNameNoNIL)
        _dir := fsCreateDir(dir, _singleFolder)
        var j int = 0
        j: for {
          var url string = fmt.Sprintf("https://uploads.mangadex.org/data/%s/%s", mangaChapterImages.Chapter.Hash, mangaChapterImages.Chapter.Data[j])
          var chapterImage []byte
          for {
            resp, err := http.Get(url)
            if err != nil {
              fmt.Println("Request error. Retrying.")
              continue
            }
            defer resp.Body.Close()
            res, err := ioutil.ReadAll(resp.Body)
            if err != nil {
              fmt.Println("Request error. Retrying.")
            } else {
              chapterImage = res
              break
            }
          }

          fsCreateFile(mangaChapterImages.Chapter.Data[j], _dir, j + 1, chapterImage, false, "")
          if j < len(mangaChapterImages.Chapter.Data) - 1 {
            j++
          } else {
            break j
          }
        }
      }
    }
    if i < len(_mangaChapters.Data) - 1 {
      i++
    } else {
      break i
    }
  }
}
